### ISSUE TITLE
debug assert when appending request headers

### ISSUE THREAD

**Issue by snowp** [envoyproxy/envoy#3919]
```
[2018-07-20 20:46:32.953][16][critical][assert] source/common/http/header_map_impl.cc:326] assert failure: *ref_lookup_response.entry_ == nullptr.
[2018-07-20 20:46:33.897][16][critical][backtrace] bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:125] Caught Segmentation fault, suspect faulting address 0x0
[2018-07-20 20:46:33.898][16][critical][backtrace] bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:94] Backtrace thr<16> obj</lib64/libc.so.6> (If unsymbolized, use tools/stack_decode.py):
[2018-07-20 20:46:33.899][16][critical][backtrace] bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:114] thr<16> #0 0x7fe2c24b6ab7 abort
[2018-07-20 20:46:33.899][16][critical][backtrace] bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:104] thr<16> obj</usr/local/bin/envoy>
[2018-07-20 20:46:33.899][16][critical][backtrace] bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:114] thr<16> #1 0xfa6fed Envoy::Http::HeaderMapImpl::insertByKey()
[2018-07-20 20:46:33.899][16][critical][backtrace] bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:114] thr<16> #2 0xfa757d Envoy::Http::HeaderMapImpl::addReferenceKey()
[2018-07-20 20:46:33.899][16][critical][backtrace] bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:114] thr<16> #3 0xcd9bd5 Envoy::Router::HeaderParser::evaluateHeaders()
[2018-07-20 20:46:33.899][16][critical][backtrace] bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:114] thr<16> #4 0xcba09a Envoy::Router::RouteEntryImplBase::finalizeRequestHeaders()
[2018-07-20 20:46:33.899][16][critical][backtrace] bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:114] thr<16> #5 0xc140f3 Envoy::Router::Filter::decodeHeaders()
[2018-07-20 20:46:33.899][16][critical][backtrace] bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:114] thr<16> #6 0xc705e0 Envoy::Http::ConnectionManagerImpl::ActiveStreamDecoderFilter::decodeHeaders()
[2018-07-20 20:46:33.899][16][critical][backtrace] bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:114] thr<16> #7 0xc67527 Envoy::Http::ConnectionManagerImpl::ActiveStream::decodeHeaders()
[2018-07-20 20:46:33.899][16][critical][backtrace] bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:114] thr<16> #8 0xc66c7c Envoy::Http::ConnectionManagerImpl::ActiveStream::decodeHeaders()
[2018-07-20 20:46:33.899][16][critical][backtrace] bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:114] thr<16> #9 0xcf8de8 Envoy::Http::Http1::ServerConnectionImpl::onMessageComplete()
[2018-07-20 20:46:33.899][16][critical][backtrace] bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:114] thr<16> #10 0xcf77ec Envoy::Http::Http1::ConnectionImpl::onMessageCompleteBase()
[2018-07-20 20:46:33.899][16][critical][backtrace] bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:114] thr<16> #11 0xcf63c5 Envoy::Http::Http1::ConnectionImpl::{lambda()#7}::operator()()
[2018-07-20 20:46:33.899][16][critical][backtrace] bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:114] thr<16> #12 0xcf63ea Envoy::Http::Http1::ConnectionImpl::{lambda()#7}::_FUN()
[2018-07-20 20:46:33.899][16][critical][backtrace] bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:114] thr<16> #13 0xd033a6 http_parser_execute
[2018-07-20 20:46:33.899][16][critical][backtrace] bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:114] thr<16> #14 0xcf72c5 Envoy::Http::Http1::ConnectionImpl::dispatchSlice()
[2018-07-20 20:46:33.899][16][critical][backtrace] bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:114] thr<16> #15 0xcf7199 Envoy::Http::Http1::ConnectionImpl::dispatch()
[2018-07-20 20:46:33.899][16][critical][backtrace] bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:114] thr<16> #16 0xc63fdc Envoy::Http::ConnectionManagerImpl::onData()
[2018-07-20 20:46:33.899][16][critical][backtrace] bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:114] thr<16> #17 0x968ea0 Envoy::Network::FilterManagerImpl::onContinueReading()
[2018-07-20 20:46:33.899][16][critical][backtrace] bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:114] thr<16> #18 0x968fa5 Envoy::Network::FilterManagerImpl::onRead()
[2018-07-20 20:46:33.899][16][critical][backtrace] bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:114] thr<16> #19 0x95fc6d Envoy::Network::ConnectionImpl::onRead()
[2018-07-20 20:46:33.899][16][critical][backtrace] bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:114] thr<16> #20 0x9611b4 Envoy::Network::ConnectionImpl::onReadReady()
[2018-07-20 20:46:33.899][16][critical][backtrace] bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:114] thr<16> #21 0x960fa3 Envoy::Network::ConnectionImpl::onFileEvent()
[2018-07-20 20:46:33.899][16][critical][backtrace] bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:114] thr<16> #22 0x95df1c Envoy::Network::ConnectionImpl::ConnectionImpl()::{lambda()#3}::operator()()
[2018-07-20 20:46:33.899][16][critical][backtrace] bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:114] thr<16> #23 0x962c53 std::_Function_handler<>::_M_invoke()
[2018-07-20 20:46:33.899][16][critical][backtrace] bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:114] thr<16> #24 0x954e8b std::function<>::operator()()
[2018-07-20 20:46:33.899][16][critical][backtrace] bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:114] thr<16> #25 0x953e97 Envoy::Event::FileEventImpl::assignEvents()::{lambda()#1}::operator()()
[2018-07-20 20:46:33.899][16][critical][backtrace] bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:114] thr<16> #26 0x953f01 Envoy::Event::FileEventImpl::assignEvents()::{lambda()#1}::_FUN()
[2018-07-20 20:46:33.899][16][critical][backtrace] bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:114] thr<16> #27 0xf1c6b1 event_process_active_single_queue.isra.29
[2018-07-20 20:46:33.899][16][critical][backtrace] bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:114] thr<16> #28 0xf1ce0e event_base_loop
[2018-07-20 20:46:33.899][16][critical][backtrace] bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:114] thr<16> #29 0x94ea59 Envoy::Event::DispatcherImpl::run()
[2018-07-20 20:46:33.900][16][critical][backtrace] bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:114] thr<16> #30 0x93e87c Envoy::Server::WorkerImpl::threadRoutine()
[2018-07-20 20:46:33.900][16][critical][backtrace] bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:114] thr<16> #31 0x93e1bf Envoy::Server::WorkerImpl::start()::{lambda()#1}::operator()()
[2018-07-20 20:46:33.900][16][critical][backtrace] bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:114] thr<16> #32 0x93ef72 std::_Function_handler<>::_M_invoke()
[2018-07-20 20:46:33.900][16][critical][backtrace] bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:114] thr<16> #33 0x4f7615 std::function<>::operator()()
[2018-07-20 20:46:33.900][16][critical][backtrace] bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:114] thr<16> #34 0xf75387 Envoy::Thread::Thread::Thread()::{lambda()#1}::operator()()
[2018-07-20 20:46:33.900][16][critical][backtrace] bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:114] thr<16> #35 0xf753ac Envoy::Thread::Thread::Thread()::{lambda()#1}::_FUN()
[2018-07-20 20:46:33.900][16][critical][backtrace] bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:104] thr<16> obj</lib64/libpthread.so.0>
[2018-07-20 20:46:33.900][16][critical][backtrace] bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:114] thr<16> #36 0x7fe2c2b55e24 start_thread
[2018-07-20 20:46:33.900][16][critical][backtrace] bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:104] thr<16> obj</lib64/libc.so.6>
[2018-07-20 20:46:33.900][16][critical][backtrace] bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:114] thr<16> #37 0x7fe2c257dbac clone
[2018-07-20 20:46:33.900][16][critical][backtrace] bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:121] end backtrace thread 16
```

this went away after setting `append: false` on `request_headers_to_add`

From looking at the code it seems like the append flow ends up calling a method that asserts that there are no existing headers. 

This is running on `c92a3017017dfa31241dec13dc6a1479090318d0/1.8.0-dev/Modified/DEBUG`

**Comment by dio** [comment]
It seems because of the header to be added (with append = true/default) is included in the `StaticLookupTable`.

My proposal will be skipping adding an `request_headers_to_add` entry (HeaderValueOption) when the entry is "static header" and `append = true`.

https://github.com/dio/envoy/commit/efe1e93cfda1cef2f9519035cc6e1a5bf729274e

**Comment by htuch** [comment]
Yeah, I bumped into this recently while writing a fuzzer for `HeaderMapImpl`. I think we should support this for consistency of the `request_headers_to_add` etc. interface. @alyssawilk

**Comment by alyssawilk** [comment]
I find the behavior of tweaking headers in the StaticLookupTable vs not super confusing - I'm glad the assert caught the fact we were dropping headers on the floor and I'd like to see a better workaround than we have today.

addReferenceKey has never handled append if there's a preexisting key.  I considered fixing that when I fixed the other append functions to work better with inline headers (#3560) but my concern was if anyone really relied on the key being a reference that'd break the contract with them.  Basically if you added "x-foo-header: bar" and then changed the backing store of x-foo-header that might not work "as expected".  That said I'd be happy to disallow that behavior as risky anyway I just think we should make it clear what's going on.  Maybe we can just change the name to indicate the function is using a static const backing store so we should avoid copies where we can but it's Ok to not use the literal string in the header entry?

**Comment by mattklein123** [comment]
> I find the behavior of tweaking headers in the StaticLookupTable vs not super confusing - I'm glad the assert caught the fact we were dropping headers on the floor and I'd like to see a better workaround than we have today.

+1 it would be great to continue to clean this up. 

> Maybe we can just change the name to indicate the function is using a static const backing store so we should avoid copies where we can but it's Ok to not use the literal string in the header entry?

Sorry can you rephrase? Didn't fully grok.

**Comment by htuch** [comment]
This showed up again under fuzzing at https://oss-fuzz.com/v2/testcase-detail/5107723602493440?noredirect=1. 

@alyssawilk (in the absence of @mattklein123); wondering if we can brainstorm some reasonable solution here. Can we just use https://cs.corp.google.com/piper///depot/google3/third_party/envoy/src/source/common/http/header_map_impl.cc?rcl=209298018&g=0&l=267 to append to the existing entry?

I notice your objection above, but I think folks should not be holding a reference to the naked string, they should instead always go via the `HeaderEntry` (or maybe `HeaderString`). 

I'm willing to take this on if we have a design consensus.

**Comment by alyssawilk** [comment]
Sure, I totally back changing all the header ByReference functions to comment that the "ByReference" is best effort, and that underlying string may be copied for multi-value/multi-line headers.  I think we'd have to look carefully at the comments for how the existing API works and make sure it's all up to date, and consider it a super-high-risk change and have Lyft (and any others that opt-in) canary carefully to make sure nothing breaks.  We know from a small bug in some of the earlier "append to inline headers" fixes that existing Envoy users are dropping some headers on the floor to the existing implementation, so the change will likely cause some changes to what upstream sees.

I think we have to make some change of this sort because it's entirely possible we're failing to proxy existing headers.  These will be changes for the better, just potentially dangerous changes for the better :-)

@ccaraman @junr03

**Comment by htuch** [comment]
OK, I will take this one on.
